src/imitation/algorithms/adversarial.py

```diff
@@ -157,33 +156,79 @@ def _log_context(self, name):
     else:
       return contextlib.nullcontext()
 
-  def train_disc(self, n_steps=10, **kwargs):
+  def train_disc(self, total_timesteps: int, **kwargs):
     """Trains the discriminator to minimize classification cross-entropy.
 
+    Every discriminator update, uses `self.batch_size` transition samples from
+    each of the expert and generator replay buffers.
+
     Args:
-        n_steps (int): The number of training steps.
-        gen_obs (np.ndarray): See `_build_disc_feed_dict`.
-        gen_acts (np.ndarray): See `_build_disc_feed_dict`.
-        gen_next_obs (np.ndarray): See `_build_disc_feed_dict`.
+      total_timesteps: The number of transitions to sample from self.venv_train.
     """
+    n_epochs = total_timesteps // self.batch_size
+    assert n_epochs >= 1, "should have at least one update"
     with self._log_context("disc"):
-      for _ in range(n_steps):
-        fd = self._build_disc_feed_dict(**kwargs)
+      for _ in tqdm(range(n_epochs), desc="Adversarial train discriminator"):
+        fd = self._build_disc_feed_dict()
         step, _ = self._sess.run([self._global_step, self._disc_train_op],
                                  feed_dict=fd)
         if self._init_tensorboard and step % 20 == 0:
           self._summarize(fd, step)
 
-  def train_gen(self, n_steps=10000):
+  def train_disc_on_samples(self, **kwargs):
+    """Trains the discriminator to minimize classification cross-entropy.
+
+    Uses the provided samples to perform a single discriminator update.
+
+    Args:
+      gen_obs (np.ndarray): See `_build_disc_feed_dict`.
+      gen_acts (np.ndarray): See `_build_disc_feed_dict`.
+      gen_next_obs (np.ndarray): See `_build_disc_feed_dict`.
+    """
+    with self._log_context("disc"):
+      fd = self._build_disc_feed_dict(**kwargs)
+      step, _ = self._sess.run([self._global_step, self._disc_train_op],
+                               feed_dict=fd)
+      if self._init_tensorboard and step % 20 == 0:
+        self._summarize(fd, step)
+
+  def train_gen(self, total_timesteps: int, callback=None):
+    """Trains the generator (via PPO2) to maximize the discriminator loss.
+
+    Args:
+      total_timesteps: The number of transitions to sample from self.venv_train.
```

Comment:
> Note these are from the same environment (I think when we spoke last you indicated they were separate environments, but I may have been misremembering).

I made them separate environments on my hp tuning branch. Will make a PR for this fix later.

> I wonder if it'd be better to just record the timesteps taken during self.gen_policy.learn though? This would be more efficient

I suppose the easiest way to retain the rollouts is to use a callback that reads `locals['obs'], locals['actions']` and loads them into a buffer. This would assume PPO2 is the underlying RL algorithm though (note that the `batch_size` => `n_steps` logic in the config file also assumes PPO2).

Do you think that environment steps are a large chunk of run time? (I haven't thought about this before).

Reply:
I'd guess collecting rollouts is the majority of the time taken by the algorithm, but haven't profiled it. Will be environment dependent obviously, but note that environment rollouts also involve policy inference, which is naturally slow and hard to parallelize.

Reply:
As you might have seen already, I collected rollouts via a wrapper in #144

Reply:
:+1: